### ISSUE TITLE
Support `IN CLUSTER` for more commands.

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#15195]
### Feature request

`CREATE MATERIALIZED VIEW` supports an `IN CLUSTER` modifier that indicates the cluster the materialized view should be installed in. Other `CREATE` commands, and also `SELECT` and `SUBSCRIBE` could also take these modifiers to more clearly indicate (and record) the target cluster. This came up as potentially being best practices for recording queries and sharing among teams, to make it clear which cluster was meant to serve certain classes of queries.

It looks like `CREATE INDEX` does support this. I'm not certain there are other `CREATE` commands to consider, but might as well review them all!

**Comment by ggnall** [comment]
This is in our private feature tracker too (for selects): https://github.com/MaterializeInc/product/issues/202. And yes worth looking at anything that could be scoped to a cluster

**Comment by benesch** [comment]
Yep, all `CREATE` commands that create objects in clusters should already support `IN CLUSTER` modifiers. I believe the exhaustive list is, in fact, `CREATE INDEX` and `CREATE MATERIALIZED VIEW`.

However, I'm reticent to extend the syntax of any DML commands (`SELECT`, `INSERT`, `DELETE`, `UPDATE`, `SUBSCRIBE`) to support overriding parameters on a per query basis. As far as I know there is no prior art for this in any database. (If there _is_, it's a much more compelling ask, and we can follow said other database's lead.)

At some level, having to ensure that you've set the `cluster` variable correctly isn't that different from having to ensure that you've set your `database` and `search_path` variables correctly. These are all variables that are typically somewhat hidden from view that substantially impact the meaning of your SQL queries. I don't want to rush to solve this with a new, nonstandard extension to SQL.

For one, I think we should explore options like making the CLI display the current cluster more prominently. This is something that could hopefully be handled by an `mz`—`psql` integration. And in our web SQL editor we of course have full flexibility.

Also, IMO there is already an acceptable way to set a variable for the duration of a single transaction only:

```
BEGIN;
SET LOCAL cluster = <transaction-scoped-value>;
SELECT ...;
COMMIT;
```

I'm not actually entirely sure we support `SET LOCAL`, but if we don't we should, because it's in PostgreSQL!

**Comment by he-la** [comment]
How about extending the existing `OPTION` syntax for DML commands to take a `cluster` option?

@benesch I agree that setting the right cluster is _in principle_ similar to connecting to the right db. In practice however I feel like materialize encourages applications to use multiple clusters much more than traditional databases encourage you to use separate logical or physical dbs for the same application (tbd on this, but that's where I see my usage of materialize heading). If my application code gets confused about which cluster it's talking to, I might not even notice during testing since the query will still run! It'll just be much slower once running on production-scale data.

Also as @frankmcsherry mentioned, when sharing code in the team, I'm very likely to have DDL and DML that needs to run on various different clusters, while still logically corresponding to the same application (e.g. having one cluster compute `materialized` views and the other cluster hold indices into that). Annotating DML commands with the intended target cluster could save a lot of comments like `-- run this in cluster xyz`.

I think it's generally good practice to allow applications to retain as little state as possible to avoid bugs, and at least the `SELECT` statement [already has syntax to support this usecase](https://materialize.com/docs/unstable/sql/select/#using-query-hints). The downside is that it's somewhat inconsistent with the `CREATE ... IN CLUSTER`.

**Comment by parker-timmerman** [comment]
I have bandwidth to work on this now, so I was hoping to draw some consensus and hammer out a proposed design for this issue!

### Goal
Just to make sure we're all aligned, as far as I understand, the goal of this specific issue is:

"Provide users a way to specify what cluster a statement will execute on, within the statement itself."

Related issues:
* https://github.com/MaterializeInc/product/issues/202
* https://github.com/MaterializeInc/materialize/issues/15650

### Current State of the World 🌎 
Today users can specify an active cluster with `SET cluster = <cluster-name>`, and they can make it "local" to a statement by setting it as part of a transaction. e.g.

```
BEGIN;
SET LOCAL cluster = bar;
SELECT * FROM foo;
COMMIT;
```
### Proposals

- [ ] TODO, select a proposal.

#### 1. Further support the `IN CLUSTER` modifier
As originally described in the issue, today we support specifying what cluster to create a view or materialized view on with the `IN CLUSTER` modifier, e.g. `CREATE MATERIALIZED VIEW ... IN CLUSTER`. 

The **proposal** is to support this `IN CLUSTER` modifier for other statement types so a user could write a query like:
```
SELECT * FROM foo IN CLUSTER bar;
```
An argument against this proposal is that clusters are not the only variable that effects how a statement is run. Others variables are the `database`, `search_path`, or as mentioned in https://github.com/MaterializeInc/materialize/pull/17836#discussion_r1133183263 a specific replica within a cluster. If we support an `IN CLUSTER` syntax, then would we also need `IN DATABASE`, `IN SEARCH_PATH`, or `IN REPLICA`?

#### 2. Support a `SETTING ( <name> = <value>, ... )` syntax
Another **proposal** is to support a new `SETTING` clause which would allow users to specify any variable that `SET` already supports, but it makes the value for that variable, local to the specific statement. For example users could write the query:
```
SELECT * FROM foo SETTING (cluster = bar);
```
this would allow us to more easily scale to support other variables like `database`, `search_path`, etc. Users could write a query like:
```
SELECT * FROM foo SETTING (cluster = bar, replica = baz, database = bag);
```

An argument against this proposal though, is supporting the new `SETTINGS` clause could be fairly tricky to implement in our parser.

### Do we need this feature? 🤔 
Both of the above proposals require adding a SQL extension, which is possibly cumbersome from a maintenance and education perspective. 

- [ ] Is the user experience today poor enough to warrant adding a SQL extension?

**Comment by jkosh44** [comment]
> One design consideration with the `SETTING` approach is how to handle the following query:
> 
> ```
> CREATE MATERIALIZED VIEW mv AS SELECT * FROM foo SETTING (cluster = foo) IN CLUSTER bar;
> ```

I just read the previous comments. I think we can actually just not accept the `settings_clause` in `CREATE MATERIALIZED VIEW` statements.

> So, all that said, I'd like to avoid rushing into adding a SQL extension for this. In the meantime I'm hoping we can get a lot of mileage out of encouraging folks to write out the two separate statements when setting the right cluster is critical.

One thing that's not clear to me is, what has changed since @benesch has made this statement that is causing us to revisit this?

> However, I'm reticent to extend the syntax of any DML commands (SELECT, INSERT, DELETE, UPDATE, SUBSCRIBE) to support overriding parameters on a per query basis. As far as I know there is no prior art for this in any database. (If there is, it's a much more compelling ask, and we can follow said other database's lead.)

@benesch another thing that's not clear to me is, why does this same argument not apply to DDL commands? One argument for the `IN CLUSTER` syntax over the `SETTING (...)` syntax is for consistency. We've already added `IN CLUSTER` to specify the cluster to use for DDL queries, so we may as well use the existing syntax for DMLs instead of inventing something new. Otherwise we'll have different ways of accomplishing the same thing depending on the context.

Another argument for `IN CLUSTER` over `SETTING (...)` is that I feel much more comfortable extending SQL for Materialize specific concepts compared to general SQL concepts. We never have to worry about an `IN CLUSTER` being added to the standard/PostgreSQL, but there's a non-zero (though maybe small) chance that the standard/PostgreSQL may add syntax for setting params on a per-query basis.

**Comment by benesch** [comment]
> > One design consideration with the `SETTING` approach is how to handle the following query:
> > ```
> > CREATE MATERIALIZED VIEW mv AS SELECT * FROM foo SETTING (cluster = foo) IN CLUSTER bar;
> > ```
> 
> I just read the previous comments. I think we can actually just not accept the `settings_clause` in `CREATE MATERIALIZED VIEW` statements.

Yes, that was exactly the idea! The `SETTING` clause is something that can get attached to the end of a DML statement, but not to the end of a DDL statement.

> > However, I'm reticent to extend the syntax of any DML commands (SELECT, INSERT, DELETE, UPDATE, SUBSCRIBE) to support overriding parameters on a per query basis. As far as I know there is no prior art for this in any database. (If there is, it's a much more compelling ask, and we can follow said other database's lead.)
> 
> @benesch another thing that's not clear to me is, why does this same argument not apply to DDL commands? One argument for the `IN CLUSTER` syntax over the `SETTING (...)` syntax is for consistency. We've already added `IN CLUSTER` to specify the cluster to use for DDL queries, so we may as well use the existing syntax for DMLs instead of inventing something new. Otherwise we'll have different ways of accomplishing the same thing depending on the context.
> 
> Another argument for `IN CLUSTER` over `SETTING (...)` is that I feel much more comfortable extending SQL for Materialize specific concepts compared to general SQL concepts. We never have to worry about an `IN CLUSTER` being added to the standard/PostgreSQL, but there's a non-zero (though maybe small) chance that the standard/PostgreSQL may add syntax for setting session params on a per-query basis.

The design of `IN CLUSTER` in DDL statements didn't go through a whole lot of review. So it's possible it's not an ideal basis for comparison. I certainly take your point about the importance of consistency with what we've already shipped, though.

The question is really one of where do we want consistency. Do we want a consistent interface for setting clusters across DDL and DML? Or do we want a consistent interface for setting arbitrary setting parameters for a given DML statement? If we introduce dedicated syntax for setting the `cluster` variable per `SELECT`, what do we do when someone wants to set the `cluster_replica` variable per `SELECT? What about `database` and `search_path`? What about something trivial like `extra_float_digits`? What I like about the `SETTING` clause is that you get _local_ consistency for overriding any configuration parameter.

Obviously having both local and global consistency is ideal, but if forced to choose I prefer local to global.

I think the strongest argument for `IN CLUSTER` rather than `SETTING` would be if we expect to ~never need to set parameters besides `cluster` and `cluster_replica`. Then the proposal would simply be:

```sql
SELECT * FROM foo IN CLUSTER bar;
SELECT * FROM foo IN CLUSTER REPLICA bar.r1;
```

That seems fine enough, if we expect to be able to keep it like that! Clusters are really special in Materialize, and it seems ok to have dedicated syntax for such a special concept.

If we later need to support `SETTING`, though, it'll be a bit awkward, as there will be all sorts of weird edge cases and multiple ways to do the same thing, e.g.:

```
SELECT * FROM foo SETTING (cluster = bar) IN CLUSTER REPLICA bar.r1;
``

**Comment by parker-timmerman** [comment]
> One thing that's not clear to me is, what has changed since @benesch has made this statement that is causing us to revisit this?

@ggnall could you comment here? Just want to confirm this feature is still a priority.


@jkosh44 @benesch thanks for the discussion, this is great! Thinking a bit more about the current use cases of `IN CLUSTER`, I don't believe we should extend it to be used with DML statements. One way to look at `IN CLUSTER` is it's a way to specify where a resource will get installed, e.g. a materialized view or index, which is different than "specifying a variable for the scope of a statement".

Because of that, I think the right approach here is to support a `SETTING` clause on both DML and DDL statements.

Supporting it on both kinds of statements gives us consistency across the product. Thinking about this from a non-cluster perspective, I can't come up with a reason why we'd let users specify variables for the scope of a `SELECT` statement but not a `CREATE` statement. Yes it does introduce different ways to specify which cluster we'll create a resource on, but that already exists today, e.g. 

* `CREATE MATERIALIZED VIEW mv AS SELECT * FROM foo IN CLUSTER bar;`
* `BEGIN; SET LOCAL cluster = bar; CREATE MATERIALIZED VIEW mv AS SELECT * FROM foo; COMMIT;`
* `SET cluster = bar; CREATE MATERIALIZED VIEW mv AS SELECT * FROM foo; SET cluster = DEFAULT;`

all have the same end result. I think just because there is some overlap in what `IN CLUSTER` and `SETTING` would be able to do, doesn't make the product experience any worse than it is today, in terms of specifying what cluster we want to create a resource on.

> Another argument for IN CLUSTER over SETTING (...) is that I feel much more comfortable extending SQL for Materialize specific concepts compared to general SQL concepts.

I totally understand the concern about extending the syntax, we don't want to paint ourselves into a corner. I know very little (nothing?) about the history of postgres or SQL and how new keywords have been introduced over time, but [`CLUSTER`](https://www.postgresql.org/docs/current/sql-cluster.html) is already used in postgres for something different than what we use it for today. So maybe the risk of extending `IN CLUSTER` is not that much different than introducing `SETTING`?

**Comment by parker-timmerman** [comment]
For posterity, `CREATE SOURCE` also supports `IN CLUSTER`, [docs](https://materialize.com/docs/sql/create-source/#sizing-a-source).

**Comment by ggnall** [comment]
At the time of writing we didn't allow sinks or sources to be scheduled on named (`IN <cluster>`) cluster, but we know do now. So I would actually say we have solid coverage on the DDL for objects that are scheduled on specific clusters.  I do not yet have a compelling case for SELECTS or DML, not that they weren't emerge.
> It looks like CREATE INDEX does support this. I'm not certain there are other CREATE commands to consider, but might as well review them all!

I actually don't think we have much urgency to build this now. If anything we have less need for it due to 1) cluster unification (above point) and 2) additional controls for setting clusters on connection (see: https://github.com/MaterializeInc/materialize/issues/14573). There is also a related problem about answering "what cluster am I in before I run this query?" which is being thought about in https://github.com/MaterializeInc/materialize/issues/18032
>One thing that's not clear to me is, what has changed since @benesch has made this statement that is causing us to revisit this?
